CSS/css_completions.py

```diff
@@ -444,78 +445,82 @@ def parse_css_data():
-        if not view.match_selector(loc, selector_scope):
-            # if the text immediately after the caret is a HTML style tag beginning, and the character before the
-            # caret matches the CSS scope, then probably the user is typing here (where | represents the caret):
-            # <style type="text/css">.test { f|</style>
-            # i.e. after a "style" HTML open tag and immediately before the closing tag.
-            # so we want to offer CSS completions here.
-            if view.match_selector(loc, 'text.html meta.tag.style.end punctuation.definition.tag.begin.html') and \
-               view.match_selector(loc - 1, selector_scope):
-                pass
-            else:
-                return []
+        # When not inside or at the end of a CSS, don’t trigger
+        if not match_selector(view, pt, selector_scope):
+            return None
 
         if not self.props:
             self.props = parse_css_data()
             self.regex = re.compile(r"([a-zA-Z-]+):\s*$")
 
-        l = []
-        if (view.match_selector(loc, prop_value_scope) or
-            # This will catch scenarios like:
-            # - .foo {font-style: |}
-            # - <style type="text/css">.foo { font-weight: b|</style>
-            view.match_selector(loc - 1, prop_value_scope)):
-
-            alt_loc = loc - len(prefix)
-            line = view.substr(sublime.Region(view.line(alt_loc).begin(), alt_loc))
+        # complete property values
+        if match_selector(view, pt, prop_value_scope):
+            line = view.substr(sublime.Region(view.line(pt).begin(), pt - len(prefix)))
 
             match = re.search(self.regex, line)
-            if match:
-                prop_name = match.group(1)
-                if prop_name in self.props:
-                    values = self.props[prop_name]
-
-                    add_semi_colon = view.substr(sublime.Region(loc, loc + 1)) != ';'
+            if not match:
+                return None
 
-                    for value in values:
-                        desc = value + "\t" + prop_name
-                        snippet = value
+            prop_name = match.group(1)
+            values = self.props.get(prop_name)
+            if not values:
+                return None
 
-                        if add_semi_colon:
-                            snippet += ";"
+            add_semi_colon = next_none_whitespace(view, pt) != ";"
 
-                        if "$1" in snippet:
-                            desc = desc.replace("$1", "")
+            items = []
+            for value in values:
+                desc = value + "\t" + prop_name
+                if "$1" in desc:
+                    desc = desc.replace("$1", "")
 
-                        l.append((desc, snippet))
+                snippet = value
+                if add_semi_colon:
+                    snippet += "$0;"
 
-                    return (l, sublime.INHIBIT_WORD_COMPLETIONS)
+                items.append((desc, snippet))
```

Comment:
Since you rewrote most of the function already:
If you split the two branches from the if-else into separate functions, you could turn them into generators. Would also make the comments redundant.

Reply:
Already thought about that, yes - as in PackageDev's settings module. Guess, will do so as there are some more tweaks to handle different kinds of edge cases might be needed.